pkg/chunk/encoding/chunk.go

```diff
@@ -246,3 +280,114 @@ func (it *indexAccessingChunkIterator) Batch(size int) Batch {
 func (it *indexAccessingChunkIterator) Err() error {
 	return it.acc.err()
 }
+
+// deletedChunkIterator helps with iterating over a partially deleted chunk.
+// It keeps index of next deleted interval that we would encounter as we keep moving forward with Scan() and Batch() calls
+// Usages should make sure deleted intervals that are passed in are sorted
+type deletedChunkIterator struct {
+	itr              Iterator
+	deletedIntervals []model.Interval
+
+	// we always move forward while iterating over data from chunks, so using nextDeletedIntervalsIndex as optimization
+	// we quickly validate whether timestamp of fetched value has reached next deleted time interval
+	nextDeletedIntervalsIndex int
+}
+
+// NewDeletedChunkIterator creates an iterator for a chunk which takes care of deleted data
+func NewDeletedChunkIterator(itr Iterator, deletedIntervals []model.Interval) Iterator {
+	return &deletedChunkIterator{
+		itr:                       itr,
+		deletedIntervals:          deletedIntervals,
+		nextDeletedIntervalsIndex: 0,
+	}
+}
+
+// Scan checks if we ran into next deleted interval, if yes then it jumps ahead and fetches data right after deleted interval that we ran into.
+func (dci *deletedChunkIterator) Scan() bool {
+	if dci.nextDeletedIntervalsIndex == -1 {
```

Comment:
Instead of tracking the index, we can reduce the array size everytime we cross an interval. I like this logic better: https://github.com/prometheus/prometheus/blob/52025bd7a9446c3178bf01dd2949d4874dd45f24/tsdb/querier.go#L1190-L1208

Reply:
Can we simplify the logic by reducing the array size by one everytime?